### PR TITLE
refactor: DeployApp accepts an interface instead of cfn primitives

### PR DESCRIPTION
**Pull request by efekarakus**
All application stacks need to implement the cloudformation.StackConfiguration
interface. By having "DeployApp" accept that interface we can have the
method deploy any type of application.

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Comment by efekarakus** [comment]
Ran e2e tests and they pass ✅
```
Ginkgo ran 5 suites in 1h34m47.3868884s
Test Suite Passed
```

File: internal/pkg/cli/app_deploy.go [hunk 1]
```diff
@@ -362,79 +358,106 @@ func (o *appDeployOpts) getAppDockerfilePath() (string, error) {
...
+	url, err := o.s3Service.PutArtifact(resources.S3Bucket, fmt.Sprintf(archer.AddonsCfnTemplateNameFormat, o.AppName), reader)
 	if err != nil {
 		return "", fmt.Errorf("put addons artifact to bucket %s: %w", resources.S3Bucket, err)
 	}
 	return url, nil
 }
 
-func (o *appDeployOpts) deployAppStack(appTemplate *bytes.Buffer, addonsURL string) error {
-	id, err := uuid.NewRandom()
+func (o *appDeployOpts) manifest() (interface{}, error) {
```
Comment (iamhopaul123) [inline internal/pkg/cli/app_deploy.go:387]: Is this function only to get the type of the application 🤔feel like we can store the app type in param store.

File: internal/pkg/deploy/cloudformation/app.go [hunk 2]
```diff
@@ -11,15 +11,19 @@ import (
...
-// creating a stack versus updating a stack.
-func (cf CloudFormation) DeployApp(template, stackName, changeSetName, cfExecutionRole string, tags map[string]string, parameters map[string]string) error {
-	stack := cloudformation.NewStack(stackName, template,
-		cloudformation.WithParameters(parameters),
-		cloudformation.WithTags(tags),
-		cloudformation.WithRoleARN(cfExecutionRole))
+// DeployApp deploys an application stack and waits until the deployment is done.
+// If the application doesn't exist, then it creates the stack.
+// If the application already exists, it updates the stack.
+func (cf CloudFormation) DeployApp(conf StackConfiguration, opts ...cloudformation.StackOption) error {
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/app.go:17]: nice!

File: internal/pkg/cli/app_deploy.go (hunk 1 above)
Comment (efekarakus) [reply internal/pkg/cli/app_deploy.go:387]: it's fetching the whole manifest object not just the type :) the whole object is needed for the `stack.NewLoadBalancedWebApp` constructor

File: internal/pkg/cli/app_deploy.go [hunk 3]
```diff
@@ -49,15 +48,16 @@ type appDeployVars struct {
...
-	s3Service          artifactPutter
-	runner             runner
-	appPackageCfClient projectResourcesGetter
-	appDeployCfClient  cloudformation.CloudFormation
-	sessProvider       sessionProvider
+	projectService   projectService
+	workspaceService wsAppReader
+	ecrService       ecrService
+	dockerService    dockerService
+	s3Service        artifactPutter
```
Comment (kohidave) [inline internal/pkg/cli/app_deploy.go:55]: just a tiny thing - can we call this artifactUploader or artifactStorer?

Comment (efekarakus) [reply internal/pkg/cli/app_deploy.go:55]: I'll incorporate this in the next PR!

**Pull request merged**